### DISCUSSION TITLE
Inconsistent behavior on many side of many-to-one relationship on new object

### DISCUSSION THREAD

**Discussion by ktchen14** [sqlalchemy/sqlalchemy#11486] [Usage Questions]
I'm trying to use `lazy="raise_on_sql"` to detect when relationships are being loaded unexpectedly. One case seemed strange to me; I've made a minimal reproduction here:

```python
from sqlalchemy import ForeignKey, Identity, create_engine
from sqlalchemy.orm import (
    DeclarativeBase, Mapped, Session, mapped_column, relationship
)

engine = create_engine("postgresql+psycopg:///")


class Record(DeclarativeBase):
    pass


class User(Record):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(Identity(), primary_key=True)
    items: Mapped[list["Item"]] = relationship(
        back_populates="user", lazy="raise_on_sql"
    )


class Item(Record):
    __tablename__ = "item"

    id: Mapped[int] = mapped_column(Identity(), primary_key=True)

    user_id: Mapped[int] = mapped_column(ForeignKey(User.id))
    user: Mapped[User] = relationship(
        back_populates="items", lazy="raise_on_sql"
    )


with Session(engine, expire_on_commit=False) as session:
    user = User(items=[])
    session.add(user)
    session.commit()
    user.items

    user = User()
    session.add(user)
    session.commit()
    user.items
```

With this SQL schema:

```sql
CREATE TABLE "user" (
  id INTEGER PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY
);

CREATE TABLE item (
  id INTEGER PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
  user_id INTEGER NOT NULL REFERENCES "user"(id)
);
```

Specifically, in this test case, the **second** `user.items` call triggers this exception:

```
sqlalchemy.exc.InvalidRequestError: 'User.items' is not available due to lazy='raise_on_sql'
```

While the first `user.items` call doesn't. The reason appears to be that, in the second case, the `items` attribute on the `user` is unloaded per `inspect(user).unloaded`. In contrast, in the first case, the collection isn't considered unloaded because of the explicit `items=[]` in the `User` constructor.

It appears to me like SQLAlchemy is essentially treating the unloaded `items` relationship attribute like a `Computed()` column. However, in that case, `eager_defaults` on a database like PostgreSQL will ensure that the column is loaded with a `RETURNING` clause. Or a client side default can be specified for the column. It doesn't look like SQLAlchemy allows me to specify an explicit default for a relationship.

This issue is similar to [this one](https://github.com/sqlalchemy/sqlalchemy/discussions/7196) but crucially differs in that I'm not doing a separate `Session.get()` after the `Session.commit()`.

Also, I tried moving the `user.items` call further up in the code. As far as I can tell, that expression correctly returns `[]` while the object is transient or pending, but as soon as it becomes persistent, the attribute access raises.

**Comment by zzzeek** [comment]
> It appears to me like SQLAlchemy is essentially treating the unloaded `items` relationship attribute like a `Computed()` column. 

I wouldn't say that.  `items` is not a column.  it's a related collection of objects that gets loaded from the `item` table, so there is no conceptual similarity here.  There is no "related collection" concept in relational databases, this is an abstraction created in terms of Python objects.   SQLAlchemy is a database-centric tool so it does not go very far in trying to create the illusion that we're using a hierarchical object database, only enough to save us the tedium of writing lots of repetitive code that can be automated.

>However, in that case, `eager_defaults` on a database like PostgreSQL will ensure that the column is loaded with a `RETURNING` clause. 

RETURNING modifies an INSERT statement, for `User` that would be `INSERT INTO user`.  There is no `RETURNING` clause here that has anything to do with rows in the `item` table, which additionally by definition can't be inserted yet because the parent row in `user` is only being inserted now.   so this has nothing to do with RETURNING or eager_defaults which only refers to local columns in the `user` table.

> Or a client side default can be specified for the column. 

this is again related to columns, not related collections which are not directly a database concept

> It doesn't look like SQLAlchemy allows me to specify an explicit default for a relationship.

Three ways to do this:

1. **in your constructor**   A "default" with a static value would not be appropriate since this is a mutable collection, so you would need to generate an empty collection on the fly:

```py
class User(Record):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(Identity(), primary_key=True)
    items: Mapped[list["Item"]] = relationship(
        back_populates="user", lazy="raise_on_sql"
    )

    def __init__(self, **kw):
        if 'items' not in kw:
            kw['items'] = []
        super().__init__(**kw)
```

2. **in the constructor, but generated automatically using dataclasses** - a more succinct way to get a "default factory" is to use dataclasses which can generate an `__init__` method for you:

```py
from sqlalchemy.orm import MappedAsDataclass

class User(MappedAsDataclass, Record):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(Identity(), primary_key=True)
    items: Mapped[list["Item"]] = relationship(
        back_populates="user", lazy="raise_on_sql", default_factory=lambda: []
    )

```

3. **using events** - you can change SQLAlchemy's behavior across the board or local to any class hierarchy using the [init](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.init) event, here using mapper inspection to locate all relationships and set them up:

```py
from sqlalchemy import event, inspect

# ...


class Record(DeclarativeBase):
    pass


@event.listens_for(Record, "init", propagate=True)
def populate_all_collections(target, args, kwargs):
    mapper = inspect(target).mapper
    for rel in mapper.relationships:
        if rel.key not in kwargs:
            kwargs[rel.key] = rel.collection_class()

# rest of program
```

> 
> Also, I tried moving the `user.items` call further up in the code. As far as I can tell, that expression correctly returns `[]` while the object is transient or pending, but as soon as it becomes persistent, the attribute access raises.

if the collection weren't addressed at all, SQLAlchemy doesn't assume that nothing else in the transaction may have affected the net value of this collection, or that there may be rows in the database that already affect the net value of this collection (if you were using a custom primaryjoin condition, for example).   if we turn off the raise_on_sql, it gives us the right answer here because it does not make this assumption:

```py
with Session(engine, expire_on_commit=False) as session:
    user = User(id=1)
    session.add(user)
    i1 = Item(user_id=1)
    session.add(i1)
    session.commit()
    assert user.items == [i1]
```

if items defaulted to an empty list then we wouldn't get the correct answer at the end.  There are other scenarios such as many-to-many collections where you might populate one side, the other side does not automatically get populated if it werent already present, things like that.   if it's not there, SQLAlchemy does not assume anything.

**Reply by ktchen14** [reply]
> I wouldn't say that. `items` is not a column. it's a related collection of objects that gets loaded from the `item` table, so there is no conceptual similarity here. There is no "related collection" concept in relational databases, this is an abstraction created in terms of Python objects. SQLAlchemy is a database-centric tool so it does not go very far in trying to create the illusion that we're using a hierarchical object database, only enough to save us the tedium of writing lots of repetitive code that can be automated.

> RETURNING modifies an INSERT statement, for `User` that would be `INSERT INTO user`. There is no `RETURNING` clause here that has anything to do with rows in the `item` table, which additionally by definition can't be inserted yet because the parent row in `user` is only being inserted now. so this has nothing to do with RETURNING or eager_defaults which only refers to local columns in the `user` table.

A relationship "feels" to me almost like a deferred column property, from a loading perspective. Obviously, a completely different strategy is required to load a relationship, but when I'm using SQLAlchemy and need to reason about when a relationship will be loaded, I tend to think of it like a deferred column property. That's why I brought up `eager_defaults`. It doesn't sound like this similarity is intentional, so please ignore my comment regarding `eager_defaults`.

> ```python
> @event.listens_for(Record, "init", propagate=True)
> def populate_all_collections(target, args, kwargs):
>     mapper = inspect(target).mapper
>     for rel in mapper.relationships:
>         if rel.key not in kwargs:
>             kwargs[rel.key] = rel.collection_class()
> 
> # rest of program
> ```

This is exactly what I was looking for. Thanks! One follow-up: would it be appropriate to use `set_committed_value()` here to set the "default" for the relationship collection?

> > Also, I tried moving the `user.items` call further up in the code. As far as I can tell, that expression correctly returns `[]` while the object is transient or pending, but as soon as it becomes persistent, the attribute access raises.
> 
> if the collection weren't addressed at all, SQLAlchemy doesn't assume that nothing else in the transaction may have affected the net value of this collection, or that there may be rows in the database that already affect the net value of this collection (if you were using a custom primaryjoin condition, for example). if we turn off the raise_on_sql, it gives us the right answer here because it does not make this assumption:
> 
> ```python
> with Session(engine, expire_on_commit=False) as session:
>     user = User(id=1)
>     session.add(user)
>     i1 = Item(user_id=1)
>     session.add(i1)
>     session.commit()
>     assert user.items == [i1]
> ```
> 
> if items defaulted to an empty list then we wouldn't get the correct answer at the end. There are other scenarios such as many-to-many collections where you might populate one side, the other side does not automatically get populated if it werent already present, things like that. if it's not there, SQLAlchemy does not assume anything.

I see. And I just noticed that the specific pending/persistent behavior for relationships is described at the bottom of this question here:

https://docs.sqlalchemy.org/en/20/faq/sessions.html#i-set-the-foo-id-attribute-on-my-instance-to-7-but-the-foo-attribute-is-still-none-shouldn-t-it-have-loaded-foo-with-id-7

Thanks for clarifying this for me. One more related question. Normally it's possible to set the loader options for an object from the database, e.g. via the `options` keyword argument to `Session.get()` or via `Select.options()`, etc. And, as far as I can tell, these loader options "live" with the object until it's reloaded with different loader options. For example, `Session.refresh()` says:

> [relationship()](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship) oriented attributes will also be immediately loaded if they were already eagerly loaded on the object, using the same eager loading strategy that they were loaded with originally

I can't find any method for "baking" loader options into a newly created object or for overriding them for an existing object short of issuing another load. Is there any way to take an existing object, and without reloading it, set the loader options that it's supposed to have over its lifetime? Thanks!

**Reply by zzzeek** [reply]
> This is exactly what I was looking for. Thanks! One follow-up: would it be appropriate to use `set_committed_value()` here to set the "default" for the relationship collection?

the object starts in the "transient" state and is not database persisted, so nothing is "committed", so no you wouldn't do this.

> 
> > > Also, I tried moving the `user.items` call further up in the code. As far as I can tell, that expression correctly returns `[]` while the object is transient or pending, but as soon as it becomes persistent, the attribute access raises.

yes that list is ephemeral unless you add something to it.  if you dont mutate that list, it's not associated with the object.   A full history of how it came to be this way is at https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#accessing-an-uninitialized-collection-attribute-on-a-transient-object-no-longer-mutates-dict .

> 
> Thanks for clarifying this for me. One more related question. Normally it's possible to set the loader options for an object from the database, e.g. via the `options` keyword argument to `Session.get()` or via `Select.options()`, etc. And, as far as I can tell, these loader options "live" with the object until it's reloaded with different loader options. For example, `Session.refresh()` says:
> 
> > [relationship()](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship) oriented attributes will also be immediately loaded if they were already eagerly loaded on the object, using the same eager loading strategy that they were loaded with originally
> 
> I can't find any method for "baking" loader options into a newly created object or for overriding them for an existing object short of issuing another load. Is there any way to take an existing object, and without reloading it, set the loader options that it's supposed to have over its lifetime? Thanks!

There's not a public API to do this directly on an object, and it would be very difficult to make this useful for end users.   It only makes sense for loaders that run on attribute access like lazy load and raise load.    You can affect these by doing a new select and include the [populate_existing](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#populate-existing) option.
